File: src/axonize/graph.py
from __future__ import annotations

from pathlib import PurePosixPath

from axonize.models import Block, BlockType, Edge, EdgeOrigin, Node, RelationType
from axonize.normalization import normalize_heading, slugify_heading


def build_nodes(blocks: list[Block]) -> list[Node]:
    return [
        Node(
            id=block.id,
            type=block.type,
            file=block.file,
            title=block.title,
            content_hash=block.content_hash,
            structure_hash=block.structure_hash,
        )
        for block in blocks
    ]


def _normalize_doc_path(path: str) -> str:
    return PurePosixPath(path).as_posix()


def _resolve_relative_path(current_file: str, candidate: str) -> str:
    base = PurePosixPath(current_file).parent
    resolved = (base / candidate).as_posix()
    return _normalize_doc_path(resolved)


def build_reference_index(blocks: list[Block]) -> dict[str, str]:
    index: dict[str, str] = {}

    section_index: dict[tuple[str, tuple[str, ...]], str] = {}
    for block in blocks:
        index[block.id] = block.id
        file_key = _normalize_doc_path(block.file)

        if block.type is BlockType.DOCUMENT:
            index[file_key] = block.id
            if file_key.endswith(".md"):
                index[file_key[:-3]] = block.id

        if block.type is BlockType.SECTION:
            section_index[(file_key, block.heading_path)] = block.id
            leaf = slugify_heading(block.heading_path[-1])
            full = "-".join(slugify_heading(part) for part in block.heading_path if part)

            index[f"{file_key}#{leaf}"] = block.id
            index[f"{file_key}#{full}"] = block.id
            index[f"{file_key[:-3]}#{leaf}"] = block.id
            index[f"{file_key[:-3]}#{full}"] = block.id

    return index


def _normalize_target(raw_target: str, source_file: str) -> tuple[str | None, str | None]:
    target = raw_target.strip()
    if not target:
        return None, None

    if target.startswith(("http://", "https://", "mailto:")):
        return None, None

    if target.startswith("#"):
        return _normalize_doc_path(source_file), target[1:]

    if "#" in target:
        path_part, anchor = target.split("#", 1)
        if not path_part:
            path_part = source_file
        elif path_part.startswith("/"):
            path_part = path_part.lstrip("/")
        else:
            path_part = _resolve_relative_path(source_file, path_part)
        return _normalize_doc_path(path_part), anchor

    path_part = target
    if path_part.startswith("/"):
        path_part = path_part.lstrip("/")
    else:
        path_part = _resolve_relative_path(source_file, path_part)
    return _normalize_doc_path(path_part), None


def resolve_target(raw_target: str, source_file: str, reference_index: dict[str, str]) -> str | None:
    if raw_target in reference_index:
        return reference_index[raw_target]

    path, anchor = _normalize_target(raw_target, source_file)
    if not path:
        return None

    candidates: list[str] = [path]
    if not path.endswith(".md"):
        candidates.append(f"{path}.md")
    if path.endswith(".md"):
        candidates.append(path[:-3])

    if anchor is None:
        for candidate in candidates:
            if candidate in reference_index:
                return reference_index[candidate]
        return None

    anchor_text = normalize_heading(anchor)
    anchor_slug = slugify_heading(anchor_text)
    for candidate in candidates:
        joined = [
            f"{candidate}#{anchor_text}",
            f"{candidate}#{anchor_text.lower()}",
            f"{candidate}#{anchor_slug}",
        ]
        for key in joined:
            if key in reference_index:
                return reference_index[key]
    return None


def _dedupe_edges(edges: list[Edge]) -> list[Edge]:
    unique: dict[tuple[str, str, str, str], Edge] = {}
    for edge in edges:
        key = (edge.source, edge.target, edge.type.value, edge.origin.value)
        existing = unique.get(key)
        if existing is None:
            unique[key] = edge
            continue
        if edge.weight > existing.weight or edge.confidence > existing.confidence:
            unique[key] = edge

    ordered = list(unique.values())
    ordered.sort(key=lambda edge: (edge.source, edge.type.value, edge.target, edge.origin.value))
    return ordered


def build_deterministic_edges(blocks: list[Block], source_files: set[str] | None = None) -> list[Edge]:
    reference_index = build_reference_index(blocks)
    block_by_file: dict[str, list[Block]] = {}
    section_by_path: dict[tuple[str, tuple[str, ...]], str] = {}
    document_by_file: dict[str, str] = {}

    for block in blocks:
        block_by_file.setdefault(block.file, []).append(block)
        if block.type is BlockType.SECTION:
            section_by_path[(block.file, block.heading_path)] = block.id
        elif block.type is BlockType.DOCUMENT:
            document_by_file[block.file] = block.id

    def includes(block: Block) -> bool:
        if source_files is None:
            return True
        return block.file in source_files

    edges: list[Edge] = []

    for block in blocks:
        if not includes(block):
            continue

        if block.type is BlockType.SECTION and len(block.heading_path) > 1:
            parent_key = (block.file, block.heading_path[:-1])
            parent = section_by_path.get(parent_key)
            if parent:
                edges.append(
                    Edge(
                        source=block.id,
                        target=parent,
                        type=RelationType.EXTENDS,
                        weight=1.0,
                        confidence=1.0,
                        origin=EdgeOrigin.DETERMINISTIC,
                    )
                )

        if block.type not in (BlockType.DOCUMENT, BlockType.SECTION):
            owner = section_by_path.get((block.file, block.heading_path))
            if owner is None:
                owner = document_by_file.get(block.file)
            if owner:
                edges.append(
                    Edge(
                        source=owner,
                        target=block.id,
                        type=RelationType.DEFINES,
                        weight=1.0,
                        confidence=1.0,
                        origin=EdgeOrigin.DETERMINISTIC,
                    )
                )

        for target_ref in block.link_targets:
            target_id = resolve_target(target_ref, block.file, reference_index)
            if not target_id:
                continue
            edges.append(
                Edge(
                    source=block.id,
                    target=target_id,
                    type=RelationType.DEPENDS_ON,
                    weight=1.0,
                    confidence=1.0,
                    origin=EdgeOrigin.DETERMINISTIC,
                )
            )

        for annotation in block.annotations:
            target_id = resolve_target(annotation.raw_target, block.file, reference_index)
            if not target_id:
                continue
            edges.append(
                Edge(
                    source=block.id,
                    target=target_id,
                    type=annotation.relation,
                    weight=1.0,
                    confidence=1.0,
                    origin=EdgeOrigin.DETERMINISTIC,
                )
            )

    return _dedupe_edges(edges)

File: desktop/main/semantic/graph.ts
import path from "node:path";
import { Block, BlockType, Edge, Node, RelationType } from "./models";
import { normalizeHeading, slugifyHeading } from "./normalization";

function normalizeDocPath(value: string): string {
  return value.split(path.sep).join("/");
}

function resolveRelativePath(currentFile: string, candidate: string): string {
  const base = path.posix.dirname(currentFile);
  return normalizeDocPath(path.posix.normalize(path.posix.join(base, candidate)));
}

export function buildNodes(blocks: readonly Block[]): Node[] {
  return blocks.map((block) => ({
    id: block.id,
    type: block.type,
    file: block.file,
    title: block.title,
    content_hash: block.contentHash,
    structure_hash: block.structureHash
  }));
}

export function buildReferenceIndex(blocks: readonly Block[]): Map<string, string> {
  const index = new Map<string, string>();

  for (const block of blocks) {
    const fileKey = normalizeDocPath(block.file);
    index.set(block.id, block.id);

    if (block.type === "document") {
      index.set(fileKey, block.id);
      if (fileKey.endsWith(".md")) {
        index.set(fileKey.slice(0, -3), block.id);
      }
    }

    if (block.type === "section" && block.headingPath.length > 0) {
      const leaf = slugifyHeading(block.headingPath[block.headingPath.length - 1]);
      const full = block.headingPath.map((segment) => slugifyHeading(segment)).filter(Boolean).join("-");
      index.set(`${fileKey}#${leaf}`, block.id);
      index.set(`${fileKey}#${full}`, block.id);
      if (fileKey.endsWith(".md")) {
        const short = fileKey.slice(0, -3);
        index.set(`${short}#${leaf}`, block.id);
        index.set(`${short}#${full}`, block.id);
      }
    }
  }

  return index;
}

function normalizeTarget(rawTarget: string, sourceFile: string): { path: string | null; anchor: string | null } {
  const target = rawTarget.trim();
  if (target.length === 0) {
    return { path: null, anchor: null };
  }

  if (target.startsWith("http://") || target.startsWith("https://") || target.startsWith("mailto:")) {
    return { path: null, anchor: null };
  }

  if (target.startsWith("#")) {
    return { path: normalizeDocPath(sourceFile), anchor: target.slice(1) };
  }

  if (target.includes("#")) {
    const [pathPartRaw, anchor] = target.split("#", 2);
    let pathPart = pathPartRaw;
    if (!pathPart) {
      pathPart = sourceFile;
    } else if (pathPart.startsWith("/")) {
      pathPart = pathPart.replace(/^\/+/, "");
    } else if (pathPart.startsWith("docs/")) {
      pathPart = pathPart;
    } else {
      pathPart = resolveRelativePath(sourceFile, pathPart);
    }
    return { path: normalizeDocPath(pathPart), anchor };
  }

  let pathPart = target;
  if (pathPart.startsWith("/")) {
    pathPart = pathPart.replace(/^\/+/, "");
  } else if (pathPart.startsWith("docs/")) {
    pathPart = pathPart;
  } else {
    pathPart = resolveRelativePath(sourceFile, pathPart);
  }
  return { path: normalizeDocPath(pathPart), anchor: null };
}

export function resolveTarget(rawTarget: string, sourceFile: string, referenceIndex: Map<string, string>): string | null {
  if (referenceIndex.has(rawTarget)) {
    return referenceIndex.get(rawTarget) ?? null;
  }

  const normalized = normalizeTarget(rawTarget, sourceFile);
  if (!normalized.path) {
    return null;
  }

  const candidates = [normalized.path];
  if (!normalized.path.endsWith(".md")) {
    candidates.push(`${normalized.path}.md`);
  }
  if (normalized.path.endsWith(".md")) {
    candidates.push(normalized.path.slice(0, -3));
  }

  if (!normalized.anchor) {
    for (const candidate of candidates) {
      if (referenceIndex.has(candidate)) {
        return referenceIndex.get(candidate) ?? null;
      }
    }
    return null;
  }

  const anchorText = normalizeHeading(normalized.anchor);
  const anchorSlug = slugifyHeading(anchorText);
  for (const candidate of candidates) {
    for (const key of [`${candidate}#${anchorText}`, `${candidate}#${anchorText.toLowerCase()}`, `${candidate}#${anchorSlug}`]) {
      if (referenceIndex.has(key)) {
        return referenceIndex.get(key) ?? null;
      }
    }
  }

  return null;
}

function dedupeEdges(edges: readonly Edge[]): Edge[] {
  const unique = new Map<string, Edge>();
  for (const edge of edges) {
    const key = `${edge.source}::${edge.target}::${edge.type}::${edge.origin}`;
    const existing = unique.get(key);
    if (!existing || edge.weight > existing.weight || edge.confidence > existing.confidence) {
      unique.set(key, edge);
    }
  }
  return [...unique.values()].sort((a, b) => {
    const aKey = `${a.source}|${a.type}|${a.target}|${a.origin}`;
    const bKey = `${b.source}|${b.type}|${b.target}|${b.origin}`;
    return aKey.localeCompare(bKey);
  });
}

export function buildDeterministicEdges(blocks: readonly Block[]): Edge[] {
  const referenceIndex = buildReferenceIndex(blocks);
  const sectionByPath = new Map<string, string>();
  const documentByFile = new Map<string, string>();

  for (const block of blocks) {
    if (block.type === "section") {
      sectionByPath.set(`${block.file}::${block.headingPath.join(" / ")}`, block.id);
    } else if (block.type === "document") {
      documentByFile.set(block.file, block.id);
    }
  }

  const edges: Edge[] = [];
  for (const block of blocks) {
    if (block.type === "section" && block.headingPath.length > 1) {
      const parentPath = block.headingPath.slice(0, -1);
      const parent = sectionByPath.get(`${block.file}::${parentPath.join(" / ")}`);
      if (parent) {
        edges.push({
          source: block.id,
          target: parent,
          type: "extends",
          weight: 1.0,
          confidence: 1.0,
          origin: "deterministic"
        });
      }
    }

    if (block.type !== "document" && block.type !== "section") {
      let owner: string | undefined;
      if (block.headingPath.length > 0) {
        owner = sectionByPath.get(`${block.file}::${block.headingPath.join(" / ")}`);
      }
      if (!owner) {
        owner = documentByFile.get(block.file);
      }
      if (owner) {
        edges.push({
          source: owner,
          target: block.id,
          type: "defines",
          weight: 1.0,
          confidence: 1.0,
          origin: "deterministic"
        });
      }
    }

    for (const targetRef of block.linkTargets) {
      const targetId = resolveTarget(targetRef, block.file, referenceIndex);
      if (!targetId) {
        continue;
      }
      edges.push({
        source: block.id,
        target: targetId,
        type: "depends_on",
        weight: 1.0,
        confidence: 1.0,
        origin: "deterministic"
      });
    }

    for (const annotation of block.annotations) {
      const targetId = resolveTarget(annotation.rawTarget, block.file, referenceIndex);
      if (!targetId) {
        continue;
      }
      edges.push({
        source: block.id,
        target: targetId,
        type: annotation.relation as RelationType,
        weight: 1.0,
        confidence: 1.0,
        origin: "deterministic"
      });
    }
  }

  return dedupeEdges(edges);
}
